src/main/java/com/google/sps/TextDetectionAPIImpl.java

```diff
@@ -0,0 +1,96 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps;
+
+import com.google.cloud.vision.v1.AnnotateImageRequest;
+import com.google.cloud.vision.v1.AnnotateImageResponse;
+import com.google.cloud.vision.v1.BatchAnnotateImagesResponse;
+import com.google.cloud.vision.v1.EntityAnnotation;
+import com.google.cloud.vision.v1.Image;
+import com.google.cloud.vision.v1.ImageAnnotatorClient;
+import com.google.sps.data.ShoppingListTextEntry;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class TextDetectionAPIImpl implements TextDetectionAPI {
+  public List<ShoppingListTextEntry> detect(byte[] imageBytes) throws PhotoDetectionException {
+    Image shoppingImage = PhotoShoppingUtil.getImageFromBytes(imageBytes);
+
+    List<AnnotateImageRequest> requests = generateShoppingImageRequest(shoppingImage);
+
+    BatchAnnotateImagesResponse response = detectTextFromImage(requests);
+
+    return parseAnnotateImageResponse(response);
+  }
+
+  /** Generates the request query to be sent to CloudVisionAPI client. */
+  private List<AnnotateImageRequest> generateShoppingImageRequest(Image shoppingImage) {
+    List<AnnotateImageRequest> requests = new ArrayList<>();
+
+    AnnotateImageRequest request =
+        AnnotateImageRequest.newBuilder()
+            .addFeatures(Constants.TEXT_DETECTION_FEATURE)
+            .setImage(shoppingImage)
+            .build();
+
+    requests.add(request);
+    return requests;
+  }
+
+  /**
+   * Sends request to cloudVisionAPI. The Cloud Vision API scans the image and returns back the
+   * text, its position and properties as the response.
+   */
+  private BatchAnnotateImagesResponse detectTextFromImage(List<AnnotateImageRequest> requests)
+      throws PhotoDetectionException {
+    ImageAnnotatorClient cloudVisionClient;
+    try {
+      cloudVisionClient = ImageAnnotatorClient.create();
+    } catch (IOException exception) {
+      throw new PhotoDetectionException(
+          "Failed to create cloudVisionClient\n" + exception.getMessage(), exception);
+    }
+    BatchAnnotateImagesResponse response = cloudVisionClient.batchAnnotateImages(requests);
+    cloudVisionClient.close();
+    return response;
+  }
+
+  /**
+   * Takes cloudVisionAPI's response and returns a list of text and their y-axis position.
+   * ToDo : The positions from annotation will be used in sentence formation algorithm to separate
+   * individual queries from the shopping list.
```

Comment:
I could imagine the x-axis position is also valuable to determine a line break, in particular, when the lines are not perfectly horizontal in a photo. The x position should only move backwards when a new line starts.

Reply:
Very valid point for a slant line Steffen. I will try and incorporate x-axis reference check too. 

Initially, I did consider taking x-axis as reference point because x-axis logic will be easier to implement (as it doesn't have to consider noise ration).
But later I went with "y-axis + height" because of one scenario. Today cloudVisionAPI sends words in the order of sentences from "top to bottom" and "left to right". But in the future, if this API changes and starts sending words randomly mixing the sentences, then x-axis based algorithm will fail; whereas "y-axis + height" algorithm will work fine. On the other hand, if API doesn't change, then x-axis reference will be a better approach.

Considering  both the points or random words, slant lines, I think that it is good to develop algorithm based on x-axis + y-axis + height; as handwritten text adds a lot of complexity to the logic. Thank you, I will keep this scenario in mind while designing the algorithm for handwritten text.